### PR TITLE
Connection lost token

### LINKED ISSUE #3324: "Directory not found" dialogue blocks jupyterlab
Excited to be on 0.30.2 (conda distribution) - Keep up the good stuff!

I've navigated into a subdirectory `a/b` with the file browser in jupyterlab. Then I used the terminal to rename `a` to `c` (accidentally initially, but it reproduces smoothly!).

A modal dialogue pops up saying "Server Connection Error", Directory not found: "a/b".
The modal reappears when clicking "OK".

If I rename `b` (the sub-subdirectory), jupyterlab shows a similar dialogue and resorts to going back to the parent directory. But not in the case renaming `a`, it renders the application unusable.

I use another terminal (via ssh) to undo the rename and jupyterlab recovers.

It would be great if jupyterlab wouldn't freeze. (Also I doubt it is actually a "Server Connection Error".)

### LINKED ISSUE #6176: extend/override connection lost for jupyterhub
In JupyterHub, there's a typical cause for "Connection Lost" (server stopped) and an action to take (go here to restart it). It would be useful to be able to customize the Connection Lost event when running with JupyterHub to point to the resolution, rather than the generic message that doesn't know how to restart the server. Does such a mechanism already exist?

cf https://github.com/jupyterhub/jupyterlab-hub/pull/84

### PR DESCRIPTION
**Pull request by ian-r-rose**
Provides a token for an action that can be taken when a connection to the server is lost. Fixes #6176. Re-fixes #3324, which had reverted.

## References

#6176 
#3324
## Code changes

Adds some new `connectionFailure` signals to service manager classes. I only added them to the ones which poll the server, which is not all of them. It feels a little asymmetric, but I mostly wanted this signal to be the result of slower polling.

## User-facing changes

No user-facing changes, except in the context of a JupyterHub deployment, where a different dialog may be provided.

## Backwards-incompatible changes

Several service managers now have a new `connectionFailure` signal on their interfaces.

### CODE REVIEW

**Comment by ian-r-rose** [comment]
Note: I have a fix I want to try to make in the filebrowser model, but was hoping to hold off on that until #6394 is resolved.

File: packages/application/src/router.ts [hunk 1]
```diff
@@ -7,158 +7,13 @@ import { URLExt } from '@jupyterlab/coreutils';
...
-  ReadonlyJSONObject,
-  Token
-} from '@phosphor/coreutils';
+import { PromiseDelegate, Token } from '@phosphor/coreutils';
 
 import { DisposableDelegate, IDisposable } from '@phosphor/disposable';
 
 import { ISignal, Signal } from '@phosphor/signaling';
 
-/* tslint:disable */
```
Comment (ian-r-rose) [inline packages/application/src/router.ts]: I moved the router token and interfaces to `application/src/token.ts` as per #6334

File: packages/application/src/tokens.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { CommandRegistry } from '@phosphor/commands';
+
+import { ServerConnection, ServiceManager } from '@jupyterlab/services';
```
Comment (ian-r-rose) [inline packages/application/src/tokens.ts]: These namespaces *may* be a bit heavy to include in `tokens.ts`. We could get around it by not allowing `IConnectionLost` to take any arguments. What do you think @vidartf?

File: packages/application/src/tokens.ts [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+/**
+ * A token for which a plugin can provide to respond to connection failures
+ * to the application server.
+ */
+export const IConnectionLost = new Token<IConnectionLost>(
+  '@jupyterlab/apputils:IConnectionLost'
+);
+
+export type IConnectionLost = (
```
Comment (ian-r-rose) [inline packages/application/src/tokens.ts]: This token interface is a function. I am not aware of anywhere else where we do this, but it seemed reasonable to me.

File: packages/application-extension/src/index.tsx [hunk 4]
```diff
@@ -741,6 +748,18 @@ const paths: JupyterFrontEndPlugin<JupyterFrontEnd.IPaths> = {
   provides: JupyterFrontEnd.IPaths
 };
 
+/**
+ * The default JupyterLab paths dictionary provider.
```
Comment (blink1073) [inline packages/application-extension/src/index.tsx]: outdated comment

File: packages/application/src/connectionlost.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { showErrorMessage } from '@jupyterlab/apputils';
+
+import { ServerConnection, ServiceManager } from '@jupyterlab/services';
+
+import { IConnectionLost } from './tokens';
+
+export const ConnectionLost: IConnectionLost = async function(
```
Comment (blink1073) [inline packages/application/src/connectionlost.ts]: docstring :smile:

File: packages/application/src/tokens.ts (hunk 3 above)
Comment (blink1073) [reply packages/application/src/tokens.ts]: I'm fine with this file as-is

File: packages/services/src/manager.ts [hunk 6]
```diff
@@ -51,6 +51,13 @@ export class ServiceManager implements ServiceManager.IManager {
     this.sessions.specsChanged.connect((_, specs) => {
       this._specsChanged.emit(specs);
     });
+
+    // Relay connection failures from the service managers that poll
+    // the server for running sessions.
+    // TODO: should we also relay connection failures from other managers?
```
Comment (blink1073) [inline packages/services/src/manager.ts]: I think these two are fine for now.

**Comment by ian-r-rose** [comment]
Okay, should be ready

File: packages/application/src/tokens.ts (hunk 2 above)
Comment (vidartf) [reply packages/application/src/tokens.ts]: As they are pure type-imports, I think the current solution works well 👍

Comment (ian-r-rose) [reply packages/application/src/tokens.ts]: `ServerConnection` and `ServiceManager` both include non-types, though they are only used as types in this file. Is our pipeline smart enough to remove them?

Comment (vidartf) [reply packages/application/src/tokens.ts]: If it works, it should be `tsc` that removes it when compiling to JS, so check the `lib` folder for the solution ;)

Comment (ian-r-rose) [reply packages/application/src/tokens.ts]: Yep, correctly removed!

**Pull request merged**